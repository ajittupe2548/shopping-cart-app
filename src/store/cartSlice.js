import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  items: [], // Array of { id, title, price, image, quantity }
  totalQuantity: 0,
  totalAmount: 0,
  isOpen: false,
};

const cartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    addToCart: (state, action) => {
      const newItem = action.payload;
      const existingItem = state.items.find(item => item.id === newItem.id);

      if (existingItem) {
        existingItem.quantity += 1;
      } else {
        state.items.push({
          ...newItem,
          quantity: 1,
        });
      }

      state.totalQuantity += 1;
      state.totalAmount += newItem.price;
    },
    removeFromCart: (state, action) => {
      const id = action.payload;
      const existingItem = state.items.find(item => item.id === id);

      if (existingItem) {
        state.totalQuantity -= existingItem.quantity;
        state.totalAmount -= existingItem.price * existingItem.quantity;
        state.items = state.items.filter(item => item.id !== id);
      }
    },
    increaseQuantity: (state, action) => {
      const id = action.payload;
      const existingItem = state.items.find(item => item.id === id);

      if (existingItem) {
        existingItem.quantity += 1;
        state.totalQuantity += 1;
        state.totalAmount += existingItem.price;
      }
    },
    decreaseQuantity: (state, action) => {
      const id = action.payload;
      const existingItem = state.items.find(item => item.id === id);

      if (existingItem && existingItem.quantity > 1) {
        existingItem.quantity -= 1;
        state.totalQuantity -= 1;
        state.totalAmount -= existingItem.price;
      }
    },
    clearCart: state => {
      state.items = [];
      state.totalQuantity = 0;
      state.totalAmount = 0;
    },
    toggleCart: state => {
      state.isOpen = !state.isOpen;
    },
    closeCart: state => {
      state.isOpen = false;
    },
  },
});

export const {
  addToCart,
  removeFromCart,
  increaseQuantity,
  decreaseQuantity,
  clearCart,
  toggleCart,
  closeCart,
} = cartSlice.actions;

export default cartSlice.reducer;
