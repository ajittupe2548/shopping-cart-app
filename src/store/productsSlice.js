import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

// Async thunk for fetching products
export const fetchProducts = createAsyncThunk(
  "products/fetchProducts",
  async (_, { rejectWithValue }) => {
    try {
      const response = await fetch("https://fakestoreapi.com/products");
      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
      const data = await response.json();
      return data;
    } catch (error) {
      return rejectWithValue(error.message);
    }
  }
);

const initialState = {
  items: [],
  status: "idle", // 'idle' | 'loading' | 'succeeded' | 'failed'
  error: null,
};

const productsSlice = createSlice({
  name: "products",
  initialState,
  reducers: {},
  extraReducers: builder => {
    builder
      .addCase(fetchProducts.pending, state => {
        state.status = "loading";
        state.error = null;
      })
      .addCase(fetchProducts.fulfilled, (state, action) => {
        state.status = "succeeded";
        state.items = action.payload;
      })
      .addCase(fetchProducts.rejected, (state, action) => {
        state.status = "failed";
        state.error = action.payload;
      });
  },
});

export default productsSlice.reducer;
