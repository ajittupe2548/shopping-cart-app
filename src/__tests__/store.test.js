import store from "../store/index";

describe("Store Configuration", () => {
  test("store is configured with correct reducers", () => {
    const state = store.getState();

    expect(state).toHaveProperty("products");
    expect(state).toHaveProperty("cart");
  });

  test("products reducer has correct initial state", () => {
    const state = store.getState();

    expect(state.products).toEqual({
      items: [],
      status: "idle",
      error: null,
    });
  });

  test("cart reducer has correct initial state", () => {
    const state = store.getState();

    expect(state.cart).toEqual({
      items: [],
      totalQuantity: 0,
      totalAmount: 0,
      isOpen: false,
    });
  });

  test("store can dispatch actions", () => {
    // Test that we can dispatch an action
    const action = { type: "cart/toggleCart" };
    expect(() => store.dispatch(action)).not.toThrow();
  });

  test("store subscription works", () => {
    let callCount = 0;
    const unsubscribe = store.subscribe(() => {
      callCount++;
    });

    store.dispatch({ type: "cart/toggleCart" });
    expect(callCount).toBe(1);

    unsubscribe();
  });

  test("matches snapshot", () => {
    const state = store.getState();
    expect(state).toMatchSnapshot();
  });
});
