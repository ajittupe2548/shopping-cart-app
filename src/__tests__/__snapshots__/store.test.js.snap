// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Store Configuration matches snapshot 1`] = `
{
  "cart": {
    "isOpen": false,
    "items": [],
    "totalAmount": 0,
    "totalQuantity": 0,
  },
  "products": {
    "error": null,
    "items": [],
    "status": "idle",
  },
}
`;
